File: app/jobs/business_lookup_job.rb
class BusinessLookupJob < ApplicationJob
  queue_as :default

  retry_on StandardError, wait: :exponentially_longer, attempts: 3

  discard_on BusinessLookupService::ProviderError do |job, exception|
    zipcode_lookup_id = job.arguments.first
    zipcode_lookup = ZipcodeLookup.find_by(id: zipcode_lookup_id)

    if zipcode_lookup
      Rails.logger.error "[BusinessLookupJob] Provider error for zipcode #{zipcode_lookup.zipcode}: #{exception.message}"
      zipcode_lookup.mark_failed!(exception)
    else
      Rails.logger.error "[BusinessLookupJob] Provider error for missing zipcode_lookup_id #{zipcode_lookup_id}: #{exception.message}"
    end
  end

  # Don't retry if zipcode lookup was deleted
  discard_on ActiveRecord::RecordNotFound do |job, exception|
    Rails.logger.error("[BusinessLookupJob] ZipcodeLookup not found: #{exception.message}")
  end

  def perform(zipcode_lookup_id)
    zipcode_lookup = ZipcodeLookup.find(zipcode_lookup_id)

    # Mark as processing
    zipcode_lookup.mark_processing!

    Rails.logger.info "[BusinessLookupJob] Starting lookup for zipcode: #{zipcode_lookup.zipcode}"

    # Perform the lookup
    service = BusinessLookupService.new(
      zipcode_lookup.zipcode,
      zipcode_lookup: zipcode_lookup
    )

    stats = service.lookup_businesses

    # Mark as completed with stats
    zipcode_lookup.mark_completed!(stats)

    Rails.logger.info "[BusinessLookupJob] Completed lookup for zipcode: #{zipcode_lookup.zipcode} - " \
                      "Found: #{stats[:found]}, Imported: #{stats[:imported]}, " \
                      "Updated: #{stats[:updated]}, Skipped: #{stats[:skipped]}"

  rescue StandardError => e
    # Guard against undefined zipcode_lookup if error occurred during find
    if defined?(zipcode_lookup) && zipcode_lookup
      Rails.logger.error "[BusinessLookupJob] Error processing zipcode #{zipcode_lookup.zipcode}: #{e.message}"
      Rails.logger.error e.backtrace.join("\n")
      zipcode_lookup.mark_failed!(e)
    else
      Rails.logger.error "[BusinessLookupJob] Error processing zipcode_lookup_id #{zipcode_lookup_id}: #{e.message}"
      Rails.logger.error e.backtrace.join("\n")
    end
    raise e
  end
end

File: laravel-app/app/Jobs/BusinessLookupJob.php
<?php

namespace App\Jobs;

use App\Models\ZipcodeLookup;
use App\Services\BusinessLookupService;
use App\Exceptions\ProviderError;
use Illuminate\Bus\Queueable;
use Illuminate\Contracts\Queue\ShouldQueue;
use Illuminate\Foundation\Bus\Dispatchable;
use Illuminate\Queue\InteractsWithQueue;
use Illuminate\Queue\SerializesModels;
use Illuminate\Support\Facades\Log;
use Exception;

class BusinessLookupJob implements ShouldQueue
{
    use Dispatchable, InteractsWithQueue, Queueable, SerializesModels;

    public $zipcodeLookupId;

    /**
     * The number of times the job may be attempted.
     *
     * @var int
     */
    public $tries = 3;

    /**
     * The number of seconds to wait before retrying.
     *
     * @var int
     */
    public $backoff = [10, 30, 90];

    /**
     * The maximum number of seconds the job can run before timing out.
     *
     * @var int
     */
    public $timeout = 300;

    /**
     * Create a new job instance.
     *
     * @param int $zipcodeLookupId
     */
    public function __construct(int $zipcodeLookupId)
    {
        $this->zipcodeLookupId = $zipcodeLookupId;
        $this->onQueue('default');
    }

    /**
     * Execute the job.
     *
     * @return void
     */
    public function handle()
    {
        $zipcodeLookup = null;

        try {
            $zipcodeLookup = ZipcodeLookup::findOrFail($this->zipcodeLookupId);

            // Mark as processing
            $zipcodeLookup->markProcessing();

            Log::info("[BusinessLookupJob] Starting lookup for zipcode: {$zipcodeLookup->zipcode}");

            // Perform the lookup
            $service = new BusinessLookupService($zipcodeLookup->zipcode, $zipcodeLookup);
            $stats = $service->lookupBusinesses();

            // Mark as completed with stats
            $zipcodeLookup->markCompleted($stats);

            Log::info("[BusinessLookupJob] Completed lookup for zipcode: {$zipcodeLookup->zipcode} - " .
                "Found: {$stats['found']}, Imported: {$stats['imported']}, " .
                "Updated: {$stats['updated']}, Skipped: {$stats['skipped']}");

        } catch (ProviderError $e) {
            if ($zipcodeLookup) {
                Log::error("[BusinessLookupJob] Provider error for zipcode {$zipcodeLookup->zipcode}: {$e->getMessage()}");
                $zipcodeLookup->markFailed($e);
            } else {
                Log::error("[BusinessLookupJob] Provider error for missing zipcode_lookup_id {$this->zipcodeLookupId}: {$e->getMessage()}");
            }
            // Don't retry on provider errors
        } catch (Exception $e) {
            if ($zipcodeLookup) {
                Log::error("[BusinessLookupJob] Error processing zipcode {$zipcodeLookup->zipcode}: {$e->getMessage()}");
                Log::error($e->getTraceAsString());
                $zipcodeLookup->markFailed($e);
            } else {
                Log::error("[BusinessLookupJob] Error processing zipcode_lookup_id {$this->zipcodeLookupId}: {$e->getMessage()}");
                Log::error($e->getTraceAsString());
            }
            throw $e;
        }
    }

    /**
     * Handle a job failure.
     *
     * @param  \Throwable  $exception
     * @return void
     */
    public function failed(\Throwable $exception)
    {
        Log::error("[BusinessLookupJob] Failed for zipcode_lookup_id {$this->zipcodeLookupId}: {$exception->getMessage()}");
    }
}
